File: 2020/Day_07/code.php
<?php

require_once __DIR__ . '/../../lib/php/utils.php';

$input = readInput(__DIR__);

// Task code
function part01(array $input): int
{
    $bagGraph = transpose(
        createBagGraph($input)
    );

    return traverse($bagGraph, 'shiny gold') - 1;
}

function part02(array $input): int
{
    $graph = createBagGraph($input);

    return countBags($graph, 'shiny gold')-1;
}

function createBagGraph(array $input): array
{
    $bagGraph = [];

    foreach ($input as $line) {
        $matches = null;
        $matched = preg_match('/(.*) bags contain (.*)/', $line, $matches);

        if (!$matched) {
            continue;
        }

        [, $outerBag, $innerBags] = $matches;
        $bagGraph[$outerBag] = [];

        $innerBagList = explode(', ', $innerBags);
        foreach ($innerBagList as $innerBag) {
            $matches = null;
            $matched = preg_match('/(\d+) (.*) (bags|bag)/', $innerBag, $matches);

            if (!$matched) {
                continue;
            }

            [, $weight, $bag] = $matches;
            $bagGraph[$outerBag][$bag] = $weight;
        }
    }

    return $bagGraph;
}

function transpose(array $graph): array
{
    $result = [];

    foreach ($graph as $outerBag => $innerBags) {
        foreach ($innerBags as $innerBag => $weight) {
            if (!array_key_exists($innerBag, $result)) {
                $result[$innerBag] = [];
            }

            $result[$innerBag][] = $outerBag;
        }
    }

    return $result;
}

function traverse(array $graph, string $node, array &$hits = []): int
{
    if (in_array($node, $hits, true)) {
        return 0;
    }

    $hits[] = $node;

    $count = 1;
    foreach ($graph[$node] ?? [] as $parent) {
        $count += traverse($graph, $parent, $hits);
    }

    return $count;
}

function countBags(array $graph, string $node): int
{
    $innerBags = $graph[$node];
    $count = 1;

    foreach ($innerBags as $innerBag => $weight) {
        $count += countBags($graph, $innerBag) * $weight;
    }

    return $count;
}

// Execute
calcExecutionTime();
$result01 = part01($input);
$result02 = part02($input);
$executionTime = calcExecutionTime();

writeln('Solution Part 1: ' . $result01);
writeln('Solution Part 2: ' . $result02);
writeln('Execution time: ' . $executionTime);

saveBenchmarkTime($executionTime, __DIR__);

// Task test
testResults(
    [326, 5635], // Expected
    [$result01, $result02], // Result
);

File: 2022/Day_01/code.go
package main

import (
	"advent-of-code/helper"
	"fmt"
	"path"
	"runtime"
	"sort"
	"strconv"
	"time"
)

func main() {
	_, f, _, _ := runtime.Caller(0)
	cwd := path.Join(path.Dir(f))

	input := helper.ReadInput(cwd, helper.NewLine)
	iValues := input.Strings()

	// Execute
	start := time.Now()
	result01 := part01(iValues)
	result02 := part02(iValues)
	executionTime := helper.ExecutionTime(time.Since(start))

	fmt.Printf("Solution Part 1: %v\n", result01)
	fmt.Printf("Solution Part 2: %v\n", result02)
	fmt.Printf("Execution time: %s\n", executionTime)

	helper.SaveBenchmarkTime(executionTime, cwd)

	// Testing
	helper.TestResults(
		[]helper.TestingValue{
			helper.TestingValue{Result: result01, Expect: 67027},
			helper.TestingValue{Result: result02, Expect: 197291},
		},
	)
}

type CalorieList []int

func NewCalorieList(input []string) CalorieList {
	list := CalorieList{}

	tmp := 0
	for _, s := range input {
		if "" == s {
			list = append(list, tmp)
			tmp = 0

			continue
		}

		i, _ := strconv.Atoi(s)
		tmp += i
	}

	return list
}

// Task code
func part01(input []string) int {
	l := NewCalorieList(input)
	sort.Ints(l)

	return l[len(l)-1]
}

func part02(input []string) int {
	l := NewCalorieList(input)
	sort.Ints(l)
	length := len(l)

	return l[length-1] + l[length-2] + l[length-3]
}
